### DISCUSSION TITLE
How does clangd implement keyword completion? such as "const", "auto", etc.

### DISCUSSION THREAD

**Discussion by Gioogs** [clangd/clangd#2132] [Q&A] [answered]
i want to add a feature in my project  that when i input the "muta", clangd can give the suggestion "mutable". like this :
![image](https://github.com/user-attachments/assets/87cf3218-8cf6-42c9-ac21-dbb81a20e64e)

**Comment by HighCommander4** [comment, marked as answer]
Completion of `mutable` is handled [here](https://searchfox.org/llvm/rev/95d2d1cba0e1428718bbdce0504292f62b212920/clang/lib/Sema/SemaCodeComplete.cpp#1904). It currently only applies to class bodies (where `mutable` can apply to the definition of a class member variable). It would need to be implemented to also apply in a lambda expression context as shown.

**Reply by Gioogs** [reply]
yes, thanks!!
Maybe add it [here](https://github.com/llvm/llvm-project/blob/ef50970204384643acca42ba4c7ca8f14865a0c2/clang/lib/Sema/SemaCodeComplete.cpp#L1830)?

**Reply by HighCommander4** [reply]
> Maybe add it [here](https://github.com/llvm/llvm-project/blob/ef50970204384643acca42ba4c7ca8f14865a0c2/clang/lib/Sema/SemaCodeComplete.cpp#L1830)?

That would offer a completion for `mutable` everywhere a type-specifier is valid, e.g. at the beginning of any statement. That's too noisy (`mutable` is not valid at the beginning of a statement).

You'd want something more specific to the context after a lambda parameter list. It's not obvious what existing function in `SemaCodeComplete` (if any) handles that. You may need to step through the code in a debugger when completion is invoked in a context and see what gets called.
